src/storage/v2/indices/vector_index.cpp

```diff
@@ -0,0 +1,147 @@
+VectorIndex::~VectorIndex() {}
+
+void VectorIndex::CreateIndex(const VectorIndexSpec &spec) {
+  // TODO(DavIvek): Take a look under https://github.com/memgraph/cmake/blob/main/vs_usearch.cpp to see how to inject
+  // custom key.
+  // TODO(DavIvek): Parametrize everything (e.g. vector_size should be dynamic).
+
+  // size and limit are important parameters for the usearch index to work properly.
+  uint64_t vector_size = spec.config["size"];
+  uint64_t limit = spec.config["limit"];
+  unum::usearch::metric_punned_t metric(vector_size, unum::usearch::metric_kind_t::l2sq_k,
+                                        unum::usearch::scalar_kind_t::f32_k);
+
+  const auto label_prop = LabelPropKey{spec.label, spec.property};
+  pimpl->index_name_to_label_prop_.emplace(spec.index_name, label_prop);
+  pimpl->index_.emplace(label_prop, mg_vector_index_t::make(metric));
+  pimpl->index_[label_prop].reserve(limit);
+
+  spdlog::trace("Created vector index " + spec.index_name);
+}
+
+void VectorIndex::AddNodeToNewIndexEntries(Vertex *vertex, std::vector<VectorIndexTuple> &keys) {
+  for (const auto &[label_prop, _] : pimpl->index_) {
+    if (utils::Contains(vertex->labels, label_prop.label()) && vertex->properties.HasProperty(label_prop.property())) {
+      keys.emplace_back(vertex, label_prop);
+    }
+  }
+}
+
+void VectorIndex::AddNodeToIndex(Vertex *vertex, const LabelPropKey &label_prop, uint64_t commit_timestamp) {
+  auto &index = pimpl->index_.at(label_prop);
+  const auto &vector_property = vertex->properties.GetProperty(label_prop.property()).ValueList();
+  std::vector<float> vector;
+  vector.reserve(vector_property.size());
+  std::transform(vector_property.begin(), vector_property.end(), std::back_inserter(vector),
+                 [](const auto &value) { return value.ValueDouble(); });
+
+  const auto key = VectorIndexKey{vertex, commit_timestamp};
+  index.add(key, vector.data());
+}
+
+std::size_t VectorIndex::Size(const std::string &index_name) {
+  const auto label_prop = pimpl->index_name_to_label_prop_.at(index_name);
+  return pimpl->index_.at(label_prop).size();
+}
+
+std::vector<std::string> VectorIndex::ListAllIndices() {
+  std::vector<std::string> indices;
+  for (const auto &[index_name, _] : pimpl->index_name_to_label_prop_) {
+    indices.push_back(index_name);
+  }
+  return indices;
+}
+
+std::vector<Vertex *> VectorIndex::Search(const std::string &index_name, uint64_t start_timestamp,
+                                          uint64_t result_set_size, const std::vector<float> &query_vector) {
+  const auto &label_prop = pimpl->index_name_to_label_prop_.at(index_name);
+  const auto &index = pimpl->index_.at(label_prop);
+
+  std::vector<Vertex *> result;
+  result.reserve(result_set_size);
+
+  auto filtering_function = [start_timestamp](const VectorIndexKey &key) {
+    // This transcation can see only nodes that were committed before the start_timestamp.
+    return key.commit_timestamp < start_timestamp;
```

Comment:
Add a comment that this works ok for creation; leave that comment as we implement a proper filter for updates as well.